native/jni/magiskhide/proc_monitor.cpp

```diff
@@ -106,6 +146,209 @@ static void hide_daemon(int pid) {
 	_exit(0);
 }
 
+/*
+ * Bionic's atoi runs through strtol() and fault-tolerence checkings.
+ * Since we don't need it, use our own implementation of atoi()
+ * for faster conversion.
+ */
+static inline int fast_atoi(const char *str) {
+	int val = 0;
+
+	while (*str)
+		val = val * 10 + (*str++ - '0');
+
+	return val;
+}
+
+// Leave /proc fd opened as we're going to read from it repeatedly
+static DIR *dfd;
+// Use unordered map with pid and namespace inode number to avoid time-consuming GC
+static unordered_map<int, uint64_t> pid_ns_map;
+// Use set for slow insertion but fast searching(which we'd encounter a lot more)
+static set<uid_t> hide_uid;
+// Treat GMS separately as we're only interested in one component
+static int gms_uid = -1;
+
+static void detect_new_processes() {
+	struct dirent *dp;
+	struct stat ns, pns;
+	int pid, ppid;
+	bool hide;
+	uid_t uid;
+	unordered_map<int, uint64_t>::const_iterator pos;
+
+	// Iterate through /proc and get a process that reads the target APK
+	rewinddir(dfd);
+	while ((dp = readdir(dfd))) {
+		if (!isdigit(dp->d_name[0]))
+			continue;
+
+		// dp->d_name is now the pid
+		pid = fast_atoi(dp->d_name);
+
+		// We're only interested in PIDs > 1000
+		if (pid <= 1000)
+			continue;
+
+		uid = get_uid(pid) % 100000; // Handle multiuser
+		if (hide_uid.find(uid) != hide_uid.end()) {
+			// Make sure our target is alive
+			if ((ppid = parse_ppid(pid)) < 0 || read_ns(ppid, &pns) || read_ns(pid, &ns))
```

Comment:
app processes are not checked whether the separation is done or not, causing MagiskHide unmount stuffs in the namespace of zygote, all apps will no longer have root access.
btw, If user uses appops to ignore an app `READ_EXTERNAL_STORAGE` op, mount namespace of its process will not be separated from zygote.

Reply:
Is that what usleep() loop with st_ino was before @topjohnwu ?
It didn't work with the rework, but I'll check the corner case you've mentioned.
Thanks.